File: seminar-05/singly_linked_list/cpp/singly_linked_list.cpp
#include <stdexcept>
#include <vector>

template<typename T>
class SinglyLinkedList {
private:
    struct Node {
        T value;
        Node* next;
        Node(const T& value, Node* next = nullptr) : value(value), next(next) {}
    };

    Node* head;
    Node* tail;
    int size_;

public:
    SinglyLinkedList() : head(nullptr), tail(nullptr), size_(0) {}

    ~SinglyLinkedList() {
        clear();
    }

    void clear() {
        Node* cur = head;
        while (cur != nullptr) {
            Node* nxt = cur->next;
            delete cur;
            cur = nxt;
        }
        head = tail = nullptr;
        size_ = 0;
    }

    bool isEmpty() const {
        return head == nullptr;
    }

    void pushFront(const T& value) {
        Node* node = new Node(value, head);
        head = node;
        if (tail == nullptr) {
            tail = node;
        }
        size_++;
    }

    void pushBack(const T& value) {
        Node* node = new Node(value);
        if (tail == nullptr) {
            head = tail = node;
        } else {
            tail->next = node;
            tail = node;
        }
        size_++;
    }

    T popFront() {
        if (isEmpty()) {
            throw std::runtime_error("pop_front from empty list");
        }
        Node* node = head;
        T value = node->value;
        head = head->next;
        delete node;
        size_--;
        if (head == nullptr) {
            tail = nullptr;
        }
        return value;
    }

    bool find(const T& value) const {
        Node* cur = head;
        while (cur != nullptr) {
            if (cur->value == value) {
                return true;
            }
            cur = cur->next;
        }
        return false;
    }

    /**
     * Удаляет первое вхождение value.
     * @return true если удалили, иначе false
     */
    bool removeFirst(const T& value) {
        Node* prev = nullptr;
        Node* cur = head;

        while (cur != nullptr) {
            if (cur->value == value) {
                if (prev == nullptr) {
                    head = cur->next;
                } else {
                    prev->next = cur->next;
                }

                if (cur == tail) {
                    tail = prev;
                }

                delete cur;
                size_--;

                if (head == nullptr) {
                    tail = nullptr;
                }
                return true;
            }

            prev = cur;
            cur = cur->next;
        }
        return false;
    }

    int size() const {
        return size_;
    }

    std::vector<T> toVector() const {
        std::vector<T> res;
        res.reserve(size_);
        Node* cur = head;
        while (cur != nullptr) {
            res.push_back(cur->value);
            cur = cur->next;
        }
        return res;
    }
};

File: seminar-05/singly_linked_list/python/singly_linked_list.py
class Node:
    """Узел односвязного списка."""

    def __init__(self, value, next=None):
        self.value = value
        self.next = next


class SinglyLinkedList:
    """Односвязный список."""

    def __init__(self):
        self.head = None
        self.tail = None
        self._size = 0

    def is_empty(self):
        return self.head is None

    def push_front(self, value):
        node = Node(value, self.head)
        self.head = node
        if self.tail is None:
            self.tail = node
        self._size += 1

    def push_back(self, value):
        node = Node(value)
        if self.tail is None:
            self.head = self.tail = node
        else:
            self.tail.next = node
            self.tail = node
        self._size += 1

    def pop_front(self):
        if self.is_empty():
            raise IndexError("pop_front from empty list")
        value = self.head.value
        self.head = self.head.next
        self._size -= 1
        if self.head is None:
            self.tail = None
        return value

    def find(self, value):
        """Возвращает True, если значение найдено."""
        cur = self.head
        while cur is not None:
            if cur.value == value:
                return True
            cur = cur.next
        return False

    def remove_first(self, value):
        """
        Удаляет первое вхождение value.
        :return: True, если удалили, иначе False
        """
        prev = None
        cur = self.head
        while cur is not None:
            if cur.value == value:
                if prev is None:
                    # удаляем голову
                    self.head = cur.next
                else:
                    prev.next = cur.next

                if cur == self.tail:
                    self.tail = prev

                self._size -= 1
                return True

            prev = cur
            cur = cur.next
        return False

    def to_list(self):
        res = []
        cur = self.head
        while cur is not None:
            res.append(cur.value)
            cur = cur.next
        return res

    def __len__(self):
        return self._size

    def __iter__(self):
        cur = self.head
        while cur is not None:
            yield cur.value
            cur = cur.next

    def __repr__(self):
        return f"SinglyLinkedList(size={self._size}, values={self.to_list()})"
